html/interaction/focus/document-level-focus-apis/document-has-system-focus.html

```diff
@@ -0,0 +1,31 @@
+<!DOCTYPE html>
+<meta charset="utf-8">
+<title>HTML Test: focus - document has system focus</title>
+<link rel="help" href="https://html.spec.whatwg.org/#has-focus-steps">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<div id="log"></div>
+<input id="test">
+<script>
+
+promise_test(async t => {
+  await new Promise(r => window.onload = r);
+  assert_true(document.hasFocus(), "Document has focus on load.");
+
+  const popup = window.open("support/popup.html", "otherwindow", "resizable");
+  let gotBlur = false;
+  window.onblur = () => gotBlur = true;
+  const msg = await new Promise(r => window.onmessage = ({data}) => r(data));
+  assert_equals(msg, "visible",
+                "Test requires popups be opened (may require harness flags)");
+  assert_true(gotBlur, "Document received blur event when popup opened");
```

Comment:
Firefox fails this on CI: https://wpt.fyi/results/html/interaction/focus/document-level-focus-apis/document-has-system-focus.html?diff&filter=ADC&run_id=5729924696506368&run_id=5675726235762688 .

I wonder if it would help to put the `window.onblur = ...` handler before the `window.open()` call? Still an interop bug, but maybe worth checking...

Reply:
@domenic When I tested it locally it worked when I hit the refresh button, but on testing it again I see it fails if I instead hit enter in the URL bar. So Firefox seems a bit unreliable here, which might impact the test harness depending on how it's launched maybe. 

I've added some code to ensure a more reliable pre-condition. LMK if that helps, as I'd rather not move the event handlers unless we have to, as that's a pet-peeve of mine.

Reply:
Still no good. https://wpt.fyi/results/html/interaction/focus/document-level-focus-apis/document-has-system-focus.html?diff&filter=ADC&run_id=5712222049271808&run_id=5679537348149248

I'm a bit surprised that you're opposed to moving the event handler setup code to before the code that triggers the event? That seems more logical generally?

Reply:
> I'm a bit surprised that you're opposed to moving the event handler setup code to before the code that triggers the event? That seems more logical generally?

@domenic I used to do that as a habit from c++. But in JS, doing so would only help if `window.open()` dispatched events synchronously.

Specs _typically_ queue a task to fire events, which means attaching event handlers after synchronous calls should work if implemented to spec. That's why I generally prefer tests written that way, to catch incorrect implementations. I admit I also find it more readable and confident.

That said, I confess that following the https://html.spec.whatwg.org/#window-open-steps I wasn't able to discern whether focus and blur events should in fact be dispatched synchronously or not. I hope not, because the tests pass in most browsers, suggesting they're not in practice.

Reply:
Yeah, I don't think there's actually any spec supporting the fact that opening a new window should switch (system) focus to that window. Much less whether that switching should be done synchronously or asynchronously. So that's why I was suggesting the more lenient version.